Prevent exception when attempting to write PaxTarEntry obtained from TarReader

carlossanlop opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/75215

A `PaxTarEntry` obtained from a `TarReader` will have an extended attributes dictionary filled with the fields that get collected by default.

When attempting to write such entry to a `TarWriter` via `WriteEntry(TarEntry)`, we collect the default extended attribute fields and store them in the dictionary. The problem is that we were using `Dictionary.Add`, which throws if the key already exists, and that is not the intended behavior.

The fix is to add the key using the indexer. That way, we _always_ write the most up-to-date value to the dictionary.

We should try to get it into 7.0 as mentioned in the issue.

carlossanlop committed:
Prevent exception when collecting extended attributes before writing entry.

carlossanlop committed:
Add tests

carlossanlop committed:
Proper method name

stephentoub commented on src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.WriteEntry.File.Tests.cs:
```diff
@@ -206,5 +208,84 @@ public void Add_SymbolicLink(TarEntryFormat format, bool createTarget)
+                        {
+                            paxEntry.GroupName = TestLongGName;
+                            paxEntry.UserName = TestLongUName;
+                        }
```
Why does PaxTarEntry need to be special-cased?  What goes wrong if you remove this and the validation of it later in the test?

carlossanlop replied:
PAX is the entry format affected by this bug. GroupName and UserName are among the extended attributes that are being affected by this fix, and those are properties that are only available in PAX and GNU. So I need to cast the object to be able to confirm their value was properly modified, because the GetNextEntry method returns the base class object TarEntry.

stephentoub replied:
Do we have a separate test that validates you can roundtrip between reader and writer without modification to the PAX entry?

carlossanlop replied:
We have some tests that verify PAX extended attributes, but not for a round trip. I'll add them here. Thanks for mentioning them.

carlossanlop replied:
Added some roundtrip verifications in the newest tests.

danmoseley commented on src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.File.Base.cs:
```diff
@@ -0,0 +1,70 @@
+                    {
+                        string testcaseName = (string)testCaseNameObj[0];
+                        if (testcaseName is "folder_file_utf8" // The legacy name field is stored in ASCII in the Ustar format
+                                         or "longpath_splitable_under255" // Folder stored under 'unarchived' runtime-assets due to NuGet not allowing very long paths
```
Maybe too late to change if it's an asset but "splittable". No need to reset for this

carlossanlop replied:
I _knew_ that word looked weird. That is my typo. But yes, it's an asset.

Jozkee commented on src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.File.Base.cs:
```diff
@@ -0,0 +1,70 @@
+
+                    foreach (object[] testCaseNameObj in GetUstarTestCaseNames())
+                    {
+                        string testcaseName = (string)testCaseNameObj[0];
```
```suggestion
                        string testCaseName = (string)testCaseNameObj[0];
```

Jozkee commented on src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.File.Base.cs:
```diff
@@ -0,0 +1,70 @@
+
+                    foreach (object[] testCaseNameObj in GetPaxAndGnuTestCaseNames())
+                    {
+                        string testCaseName = (string)testCaseNameObj[0];
+                        if (testCaseName is "longpath_splitable_under255"
+                                         or "specialfiles"
+                                         or "longpath_over255" // Folder stored under 'unarchived' runtime-assets due to NuGet not allowing very long paths
+                                         or "longfilename_over100_under255" // Same
+                                         or "file_longsymlink") // Same
```
Do you mind elaborate further why are you skipping these archives? and how do you deal with them in other tests?

carlossanlop replied:
For the first two, the reason is the same as the one I wrote a few lines above.

Generally speaking, these tests are collecting the unarchived files from the filesystem. We have to skip them because either some OS variants do not support the files in these test cases, or nuget is unable to pack them from runtime-assets and unpack them unchanged into the test data.

Other tests are not consuming the unarchived files, they are consuming the .tar or .tar.gz files directly, which we can iterate without interacting with the filesystem. But this test in particular needs to be able to compare with the filesystem, not with TarReader entries.

The skipped test cases aren't that important. What's important is the copy of an entry that comes from a TarReader into a new TarWriter, particularly in the PAX format, to verify the bug is fixed in the extended attributes.

Jozkee replied:
> this test in particular needs to be able to compare with the filesystem, not with TarReader entries.

I left a suggestion below that can help you avoid comparing against the filesystem https://github.com/dotnet/runtime/pull/75237#discussion_r966353141. Could that help removing all these `continue`s?

Jozkee commented on src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.File.Base.cs:
```diff
@@ -0,0 +1,70 @@
+                            continue;
+                        }
+
+                        if (testCaseName is not "folder_file_utf8") // The legacy name field is stored in ASCII in the GNU format
```
> // The legacy name field is stored in ASCII in the GNU format

Sorry I don't follow, how does that justify not running the test? These are archives directly taken from GNU Tar test-assets, right? How does GNU Tar handles such archives?

carlossanlop replied:
Looking at this again, I think this is a bug. The spec is not specifically indicating that the `name`, `prefix` and `linkname` fields should exclusively use ASCII.

The `.tar` asset generated with the unix `tar` tool wrote the utf8 characters without problems.

In `TarHeader.Read.cs`, we even read those 3 fields as Utf8. But we do _not_ write them as Utf8.

I have a bug fix. I tested this locally by removing the lines skipping the `folder_file_utf8` test case, and they are now passing. Will push a commit soon.

carlossanlop replied:
Update: You're currently working on this in a separate PR.

Jozkee commented on src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.WriteEntry.File.Tests.cs:
```diff
@@ -206,5 +208,84 @@ public void Add_SymbolicLink(TarEntryFormat format, bool createTarget)
+        public void WriteEntry_CopyArchive(TestTarFormat testFormat, string testCaseName, CompressionMethod compressionMethod, bool copyData)
+        {
+            TarEntryFormat entryFormat = GetEntryFormatForTestTarFormat(testFormat);
+            string pathWithExpectedFiles = GetTestCaseUnarchivedFolderPath(testCaseName);
```
Not related to this PR but why does this method uses GetCurentDirectory?

carlossanlop replied:
The dotnet process runs inside the root folder that contains the test assets. When you run the tests on the command line, you can see the folder path:

```
  ----- start Wed 07 Sep 2022 08:12:13 PM PDT =============== To repro directly: =====================================================
  pushd /home/carlos/runtime/artifacts/bin/System.Formats.Tar.Tests/Debug/net7.0-unix                   <- ***THIS FOLDER HERE***
  /home/carlos/runtime/artifacts/bin/testhost/net7.0-Linux-Debug-x64/dotnet exec --runtimeconfig System.Formats.Tar.Tests.runtimeconfig.json --depsfile System.Formats.Tar.Tests.deps.json xunit.console.dll System.Formats.Tar.Tests.dll -xml testResults.xml -nologo -notrait category=OuterLoop -notrait category=failing
```

Jozkee commented on src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.WriteEntry.File.Tests.cs:
```diff
@@ -206,5 +208,84 @@ public void Add_SymbolicLink(TarEntryFormat format, bool createTarget)
+        {
+            TarEntryFormat entryFormat = GetEntryFormatForTestTarFormat(testFormat);
+            string pathWithExpectedFiles = GetTestCaseUnarchivedFolderPath(testCaseName);
+            if (!Path.EndsInDirectorySeparator(pathWithExpectedFiles))
+            {
+                pathWithExpectedFiles = pathWithExpectedFiles + Path.DirectorySeparatorChar;
+            }
```
This already exists in PathInternal:
https://github.com/dotnet/runtime/blob/78065413b2d1b4f0ed26343567379e992a3e26ee/src/libraries/Common/src/System/IO/PathInternal.cs#L18-L19

carlossanlop replied:
Ah, thanks for the reminder.

Jozkee commented on src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.WriteEntry.File.Tests.cs:
```diff
@@ -206,5 +208,84 @@ public void Add_SymbolicLink(TarEntryFormat format, bool createTarget)
+                pathWithExpectedFiles = pathWithExpectedFiles + Path.DirectorySeparatorChar;
+            }
+
+            using Stream file = GetTarMemoryStream(compressionMethod, testFormat, testCaseName);
```
nit:
```suggestion
            using MemoryStream file = GetTarMemoryStream(compressionMethod, testFormat, testCaseName);
```
or
```suggestion
            using Stream fileMemoryStream = GetTarMemoryStream(compressionMethod, testFormat, testCaseName);
```

I thought this was a FileStream until I saw the right side.

Jozkee commented on src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.WriteEntry.File.Tests.cs:
```diff
@@ -206,5 +208,84 @@ public void Add_SymbolicLink(TarEntryFormat format, bool createTarget)
+
+            destinationArchive.Position = 0;
+
+            Dictionary<string, FileSystemInfo> expectedEntries = GetFileSystemInfosRecursive(pathWithExpectedFiles).ToDictionary(fsi =>
+                fsi.FullName.Substring(pathWithExpectedFiles.Length).Replace(Path.DirectorySeparatorChar, '/'));
```
Another way you can validate the copy is by re-iterating originArchive (along with destinationArchive) and comparing as much fields as possible against the entries in destinationReader.

carlossanlop replied:
Why not both?

Jozkee replied:
I don't see value in extracting the entries and then using them for comparison in this case, the reported error is when you copy a Tar to another without writing to the disk. IMO, the filesystem should not be involved at all.

carlossanlop replied:
But if there is a bug in the way we are reading information with the Tar APIs, then comparing the items read with the first TarReader with a second TarReader will give me the same results. The utf8 case discussed above is a good example of this.

Jozkee replied:
Then there should be a separate test that validates file system entries extracted to disk against tar archive entries.

Jozkee commented on src/libraries/System.Formats.Tar/tests/TarTestsBase.cs:
```diff
@@ -58,6 +59,8 @@ public abstract partial class TarTestsBase : FileCleanupTestBase
 
         protected const string TestGName = "group";
         protected const string TestUName = "user";
+        protected readonly string TestLongGName = new string('�', 33);
+        protected readonly string TestLongUName = new string('�', 100);
```
I think there was a policy in the repo about hardcoding non-ASCII character (ñ and ü are extended ASCII, though). @danmoseley?

Regardless, is there a reason to use them in the `WriteEntry_CopyArchive` scenario?
```suggestion
        protected readonly string TestLongGName = new string('a', 33);
        protected readonly string TestLongUName = new string('a', 100);
```

carlossanlop replied:
Ah yeah, the escaped utf8 values.

- `ä` is `\u00e4`
- `ñ` is `\u00f1`

Found here: http://daniel-hug.github.io/characters/

Jozkee replied:
The question remains, why not just use an ASCII character?

danmoseley replied:
Yeah, https://github.com/dotnet/runtime/blob/7a45201181d37329b7ab0bee541ed7c42b5d94b6/docs/coding-guidelines/coding-style.md?plain=1#L29
It's because sometimes editors garble them, or it causes notepad to add a BOM (?) 
Don't reset validation just for this -- plenty of places we're not doing this already.

carlossanlop replied:
> The question remains, why not just use an ASCII character?

We want to make sure we are able to encode/decode UTF8 chars correctly. And since ASCII is a subset of UTF8, it's covered.

Jozkee commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:
```diff
@@ -618,15 +618,12 @@ static int CountDigits(int value)
-            ExtendedAttributes.Add(PaxEaName, _name);
+            ExtendedAttributes[PaxEaName] = _name;
 
-            if (!ExtendedAttributes.ContainsKey(PaxEaMTime))
-            {
-                ExtendedAttributes.Add(PaxEaMTime, TarHelpers.GetTimestampStringFromDateTimeOffset(_mTime));
-            }
```
This behavior looks like it was intentional, is it OK that we change it from "add if it was absent" to "always set"?

carlossanlop replied:
This `CollectExtendedAttributesFromStandardFields` method does exactly what other tar tools do when creating a PAX entry:
- They always insert the values of `name` and `mtime` for all entry types, regardless if they fit or not in the standard field.
- They insert `linkname` for symlinks and hardlinks if not empty, regardless if it fits or not in the standard field.
- If `gname` or `uname` are set, they get added if their byte lengths are too large for the standard field.
- If `size` does not fit in the standard field when converted to string, it gets inserted.

If the user decides to manually change the `mtime` before writing it into the TarWriter, then the `mtime` would not be updated.

I need to add more unit tests.

carlossanlop replied:
Addressed.

Jozkee commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:
```diff
@@ -618,15 +618,12 @@ static int CountDigits(int value)
+        private void CollectExtendedAttributesFromStandardFields()
         {
-            ExtendedAttributes.Add(PaxEaName, _name);
+            ExtendedAttributes[PaxEaName] = _name;
```
What does this mean for someone building a PaxTarEntry with their user-defined extended attributes? Are we just overwriting it? is that OK? I thought you were going to perform the set only for entries that originated from a TarReader.

carlossanlop replied:
The user should change the `name`, `linkname`, `mtime`, `gname` and `uname` using the public properties we expose. 

Maybe we should have an exception that is thrown when a `PaxTarEntry` constructor that takes a dictionary has one of those 4 elements already inserted in the dictionary.

So yes, it's ok to overwrite them. If the user changes tha value of one of those 4 using the public properties, we should update the dictionary automatically, not them.

Jozkee replied:
I don't think is ok to silently overwrite them, in that case I do prefer the current `ArgumentException` to signal that those keys are reserved, or as you say, we should check up-front if the extended attributes passed to the ctor contain one of the "reserved attributes", plus, we should assert that overwrite of "reserved attributes" only occurs for `PaxTarEntry`s coming from `TarReader`.

carlossanlop replied:
As discussed on our videocall, we will store a boolean that determines if the user has modified one of the reserved fields manually via a property, in which case, we need to sync its value with the dictionary right before the entry is written to an archive.

stephentoub commented:
> We should try to get it into 7.0 as mentioned in the issue.

This hasn't moved in 10 days.  What's the plan?

carlossanlop commented:
> This hasn't moved in 10 days. What's the plan?

I had a call with @jozkee today. We have a mutual understanding now and will address the comments today.

carlossanlop committed:
Allow dictionary with reserved keys only when reading from an archive or when using conversion constructor.
Disallow dictionary with reserved keys when using the pax constructor that takes a dictionary.
Allow user to modify reserved fields via their properties and ensure they get updated in the dictionary upon writing entry to an archive.

carlossanlop committed:
Tests, sync and async

carlossanlop committed:
Address minor feedback suggestions.

carlossanlop committed:
Merge conflict fix

carlossanlop committed:
Escape test chars

carlossanlop commented:
All the apple queues are failing with an unrelated infra problem that is being investigated right now (thanks Stephen for notifying First Responders).
```
##[error]Git fetch failed with exit code: 128
```

Jozkee commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.cs:
```diff
@@ -90,13 +100,76 @@ internal sealed partial class TarHeader
         // fields have data, we store it to avoid data loss, but we don't yet expose it publicly.
         internal byte[]? _gnuUnusedBytes;
 
+        internal string Name
```
If you create an entry and then change Name using the setter, will ExtendedAttributes["path"] return the updated value? Seems to me that it won't.

Jozkee commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:
```diff
@@ -618,42 +618,52 @@ static int CountDigits(int value)
+            if (!_isPaxEaGNameSynced && !string.IsNullOrEmpty(GName))
             {
-                TryAddStringField(ExtendedAttributes, PaxEaGName, _gName, FieldLengths.GName);
+                TryAddStringField(ExtendedAttributes, PaxEaGName, GName, FieldLengths.GName);
```
`TryAddStringField` will only change the dictionary element if `Encoding.UTF8.GetByteCount(value) > maxLength` and you are signaling _isPaxEaGNameSynced regardless of that condition.

Jozkee commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:
```diff
@@ -618,42 +618,52 @@ static int CountDigits(int value)
+            if (!_isPaxEaUNameSynced && !string.IsNullOrEmpty(UName))
             {
-                TryAddStringField(ExtendedAttributes, PaxEaUName, _uName, FieldLengths.UName);
+                TryAddStringField(ExtendedAttributes, PaxEaUName, UName, FieldLengths.UName);
```
ditto

Jozkee commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:
```diff
@@ -618,42 +618,52 @@ static int CountDigits(int value)
             }
 
-            if (_size > 99_999_999)
+            Size = GetTotalDataBytesToWrite();
```
you were not doing this before, why is it needed now?

Jozkee commented:
Closing in favor of https://github.com/dotnet/runtime/pull/76404
